From 826fa7dd0c9c24795277ff0d4b24f87bd658e58d Mon Sep 17 00:00:00 2001
From: Gert Pellin <gert@pellin.be>
Date: Thu, 8 Oct 2026 20:50:07 +0200
Subject: [PATCH] [QMS-XXXX] Support Mapsforge POI files of version 3 and 4

Mapsforge POI version 3 replaced the R-tree poi_index (minLat, maxLat,
minLon, maxLon) with a plain table (lat, lon) and one index per
column. Version 4 brought the R-tree back and added a full-text index
and an index on poi_category_map.category. The other tables QMapShack
reads are the same in all three versions. OpenAndroMaps ships version
3 by default today, version 2 only in its "mapsforge_old" folder.

The queries move to CPoiFilePOIQuery and are picked by the version in
the metadata table. Versions 2 and 4 keep the existing R-tree query
per category and 0.1 degree square.

Without an R-tree every query walks all POIs of the full 0.1 degree
latitude row and looks up their categories, whatever longitude range
and how many categories are asked for. Version 2's way of loading,
one query per category and square, is therefore 10 to 40 times
slower on a version 3 file. For version 3 a single query now loads
the row segment in view for all checked categories that still miss
it. On OpenAndroMaps' Belgium file and a view around Durbuy with the
eleven "Transport Car" categories checked, loading takes 0.4 s with
version 3 against 0.2 to 0.3 s with version 2 and 4.

The writer adds a normalized_name tag (lower case, no accents) for
searching. It is no longer shown as an alternative name or as a tag.
---
 src/qmapshack/CMakeLists.txt           |   2 +
 src/qmapshack/poi/CPoiFilePOI.cpp      | 158 ++++++++++++++++---------
 src/qmapshack/poi/CPoiFilePOI.h        |  21 ++--
 src/qmapshack/poi/CPoiFilePOIQuery.cpp |  99 ++++++++++++++++
 src/qmapshack/poi/CPoiFilePOIQuery.h   |  97 +++++++++++++++
 src/qmapshack/poi/CPoiItemPOI.cpp      |   8 +-
 6 files changed, 313 insertions(+), 72 deletions(-)
 create mode 100644 src/qmapshack/poi/CPoiFilePOIQuery.cpp
 create mode 100644 src/qmapshack/poi/CPoiFilePOIQuery.h

diff --git a/src/qmapshack/CMakeLists.txt b/src/qmapshack/CMakeLists.txt
index 0e4605fb..ce0252d3 100644
--- a/src/qmapshack/CMakeLists.txt
+++ b/src/qmapshack/CMakeLists.txt
@@ -274,6 +274,7 @@ set( SRCS
     poi/CPoiDraw.cpp
     poi/CPoiFileItem.cpp
     poi/CPoiFilePOI.cpp
+    poi/CPoiFilePOIQuery.cpp
     poi/CPoiFilePOI_TagMap.cpp
     poi/CPoiIconCategory.cpp
     poi/CPoiItemPOI.cpp
@@ -669,6 +670,7 @@ set( HDRS
     poi/CPoiDraw.h
     poi/CPoiFileItem.h
     poi/CPoiFilePOI.h
+    poi/CPoiFilePOIQuery.h
     poi/CPoiIconCategory.h
     poi/CPoiItemPOI.h
     poi/CPoiList.h
diff --git a/src/qmapshack/poi/CPoiFilePOI.cpp b/src/qmapshack/poi/CPoiFilePOI.cpp
index 2ecbbbe6..bf46fbc6 100644
--- a/src/qmapshack/poi/CPoiFilePOI.cpp
+++ b/src/qmapshack/poi/CPoiFilePOI.cpp
@@ -55,9 +55,9 @@ CPoiFilePOI::CPoiFilePOI(const QString& filename, CPoiDraw* parent)
       const QString& msg = tr("Failed to open SQL database:");
       const QString& err = db.lastError().text();
       if (msgBoxShow) {
-	QMessageBox msgBox(QMessageBox::Warning, tr("SQL error..."), msg + "<br>" + err, QMessageBox::Ok, msgBoxParent);
-	msgBox.setTextFormat(Qt::RichText);
-	msgBox.exec();
+        QMessageBox msgBox(QMessageBox::Warning, tr("SQL error..."), msg + "<br>" + err, QMessageBox::Ok, msgBoxParent);
+        msgBox.setTextFormat(Qt::RichText);
+        msgBox.exec();
       } else {
         qDebug() << msg + " " + err;
       }
@@ -86,23 +86,25 @@ CPoiFilePOI::CPoiFilePOI(const QString& filename, CPoiDraw* parent)
   }
 
   if (isActivated) {
-
     QSqlQuery query("SELECT value FROM main.metadata WHERE name='version'", QSqlDatabase::database(filename + "_bbox"));
 
     const QString& version = query.next() ? query.value(0).toString() : "unknown";
-    if (version != "2") {
-      const QString& msg = tr("POI file '%1' is POI version %2. Only version 2 is supported!").arg(filename).arg(version);
-      const QString& hint = tr("See <a href='https://github.com/Maproom/qmapshack/wiki/DocGisItemsPOI'>Wiki</a> for more information.");
+    index = CPoiFilePOIQuery::indexForVersion(version);
+    if (index == CPoiFilePOIQuery::Index::eUnsupported) {
+      const QString& msg =
+          tr("POI file '%1' is POI version %2. Only versions 2, 3 and 4 are supported!").arg(filename, version);
+      const QString& hint =
+          tr("See <a href='https://github.com/Maproom/qmapshack/wiki/DocGisItemsPOI'>Wiki</a> for more information.");
       if (msgBoxShow) {
-	QMessageBox msgBox(QMessageBox::Warning, tr("POI file error..."), msg + "<br>" + hint, QMessageBox::Ok, msgBoxParent);
-	msgBox.setTextFormat(Qt::RichText);
-	msgBox.exec();
+        QMessageBox msgBox(QMessageBox::Warning, tr("POI file error..."), msg + "<br>" + hint, QMessageBox::Ok,
+                           msgBoxParent);
+        msgBox.setTextFormat(Qt::RichText);
+        msgBox.exec();
       } else {
         qDebug() << msg;
       }
       isActivated = false;
     }
-
   }
 
   // Database is no longer needed
@@ -160,7 +162,7 @@ void CPoiFilePOI::draw(IDrawContext::buffer_t& buf) {
       for (int minLatM10 = qFloor(yMin * RAD_TO_DEG * 10); minLatM10 < qCeil(yMax * RAD_TO_DEG * 10); minLatM10++) {
         if (!loadedPoisByArea.contains(categoryID) || !loadedPoisByArea[categoryID].contains(minLonM10) ||
             !loadedPoisByArea[categoryID][minLonM10].contains(minLatM10)) {
-          loadPOIsFromFile(categoryID, minLonM10, minLatM10);
+          loadPOIsFromFile(categoryID, minLonM10, minLatM10, qCeil(xMax * RAD_TO_DEG * 10) - 1);
         }
         if (poi->needsRedraw()) {
           return;
@@ -288,7 +290,7 @@ void CPoiFilePOI::findPoisIn(const QRectF& degRect, QSet<IPoiItem>& pois, QList<
         // some tiles are not loaded then
         if (!loadedPoisByArea.contains(categoryID) || !loadedPoisByArea[categoryID].contains(minLonM10) ||
             !loadedPoisByArea[categoryID][minLonM10].contains(minLatM10)) {
-          loadPOIsFromFile(categoryID, minLonM10, minLatM10);
+          loadPOIsFromFile(categoryID, minLonM10, minLatM10, qFloor(degRect.right() * 10));
         }
         for (quint64 poiFoundID : std::as_const(loadedPoisByArea[categoryID][minLonM10][minLatM10])) {
           const CPoiItemPOI& poiItemFound = loadedPois[poiFoundID];
@@ -449,7 +451,7 @@ bool CPoiFilePOI::getPoiGroupCloseBy(const QPoint& px, CPoiFilePOI::poiGroup_t&
   return false;
 }
 
-void CPoiFilePOI::loadPOIsFromFile(quint64 categoryID, int minLonM10, int minLatM10) {
+void CPoiFilePOI::loadPOIsFromFile(quint64 categoryID, int minLonM10, int minLatM10, int maxLonM10) {
   QMutexLocker lock(&mutex);
 
   // check if query is within bounds
@@ -468,54 +470,92 @@ void CPoiFilePOI::loadPOIsFromFile(quint64 categoryID, int minLonM10, int minLat
     }
   }
 
-  QSqlQuery query(QSqlDatabase::database(filename));
-  query.prepare(
-      "SELECT main.poi_index.maxLat, main.poi_index.maxLon, main.poi_index.minLat, main.poi_index.minLon, "
-      "main.poi_data.data, main.poi_data.id "
-      "FROM main.poi_data, main.poi_index "
-      "WHERE main.poi_data.id IN "
-      "("
-      "    SELECT main.poi_category_map.id "
-      "    FROM main.poi_category_map "
-      "    WHERE main.poi_category_map.id IN "
-      "    ( "
-      "        SELECT main.poi_index.id "
-      "        FROM main.poi_index "
-      "        WHERE main.poi_index.maxLat<:maxLat "
-      "        AND main.poi_index.minLat>=:minLat "
-      "        AND main.poi_index.maxLon<:maxLon "
-      "        AND main.poi_index.minLon>=:minLon "
-      "    ) "
-      "    AND main.poi_category_map.category=:categoryID "
-      ") "
-      "AND main.poi_data.id = main.poi_index.id");
-  query.bindValue(":maxLat", QString::number((minLatM10 + 1) / 10., 'f'));
-  query.bindValue(":minLat", QString::number(minLatM10 / 10., 'f'));
-  query.bindValue(":maxLon", QString::number((minLonM10 + 1) / 10., 'f'));
-  query.bindValue(":minLon", QString::number(minLonM10 / 10., 'f'));
-  query.bindValue(":categoryID", categoryID);
+  QSqlDatabase db = QSqlDatabase::database(filename);
+  if (index == CPoiFilePOIQuery::Index::eLatLon) {
+    loadPoisInRow(db, minLonM10, maxLonM10, minLatM10);
+  } else {
+    loadPoisInSquare(db, categoryID, minLonM10, minLatM10);
+  }
+  db = QSqlDatabase();
+  // Close database, as this method is called from mutiple threads.
+  QSqlDatabase::removeDatabase(filename);
+}
+
+void CPoiFilePOI::loadPoisInSquare(const QSqlDatabase& db, quint64 categoryID, int minLonM10, int minLatM10) {
+  QSqlQuery query(db);
+  query.prepare(CPoiFilePOIQuery::selectPoisInSquare());
+  CPoiFilePOIQuery::bindSquare(query, categoryID, minLonM10, minLatM10);
   query.exec();
   while (query.next()) {
-    quint64 key = query.value(eSqlColumnPoiId).toUInt();
-    const QStringList& data = query.value(eSqlColumnPoiData).toString().split("\r");
-    QString garminIcon;
-    for (const QString& tag : data) {
-      if (tagMap.contains(tag)) {
-        garminIcon = tagMap[tag].getGarminSym();
-        break;
+    addPoi(query, categoryID, minLonM10, minLatM10);
+  }
+}
+
+void CPoiFilePOI::loadPoisInRow(const QSqlDatabase& db, int minLonM10, int maxLonM10, int minLatM10) {
+  // A query costs about the same for one square as for a whole row, and for one category as for several.
+  // So load the row segment in view for all checked categories that miss its first square.
+  const int firstLonM10 = qMax(minLonM10, qFloor(bbox.left() * 10));
+  const int lastLonM10 = qMax(firstLonM10, qMin(maxLonM10, qFloor(bbox.right() * 10)));
+
+  QList<quint64> categories;
+  QSet<QPair<quint64, int>> squaresToLoad;
+  const QList<quint64>& keys = categoryActivated.keys();
+  for (quint64 categoryID : keys) {
+    if (categoryActivated[categoryID] != Qt::Checked || isLoaded(categoryID, firstLonM10, minLatM10)) {
+      continue;
+    }
+    categories << categoryID;
+    for (int lonM10 = firstLonM10; lonM10 <= lastLonM10; lonM10++) {
+      if (!isLoaded(categoryID, lonM10, minLatM10)) {
+        // mark the square as loaded, even if it has no POIs
+        loadedPoisByArea[categoryID][lonM10][minLatM10];
+        squaresToLoad.insert({categoryID, lonM10});
       }
     }
-    loadedPoisByArea[categoryID][minLonM10][minLatM10].append(key);
-    // TODO: this overwrites a POI if it already was loaded. The difference between those will be the category. Some
-    // better handling should be done
-    loadedPois[key] = CPoiItemPOI(
-        data,
-        QPointF((query.value(eSqlColumnPoiMaxLon).toDouble() + query.value(eSqlColumnPoiMinLon).toDouble()) / 2 *
-                    DEG_TO_RAD,
-                (query.value(eSqlColumnPoiMaxLat).toDouble() + query.value(eSqlColumnPoiMinLat).toDouble()) / 2 *
-                    DEG_TO_RAD),
-        key, categoryNames[categoryID], garminIcon);
   }
-  // Close database, as this method is called from mutiple threads.
-  QSqlDatabase::removeDatabase(filename);
+  if (categories.isEmpty()) {
+    return;
+  }
+
+  QSqlQuery query(db);
+  query.prepare(CPoiFilePOIQuery::selectPoisInRow(categories.size()));
+  CPoiFilePOIQuery::bindRow(query, categories, firstLonM10, lastLonM10, minLatM10);
+  query.exec();
+  while (query.next()) {
+    const quint64 categoryID = query.value(CPoiFilePOIQuery::eSqlColumnPoiCategory).toULongLong();
+    const qreal lon = query.value(CPoiFilePOIQuery::eSqlColumnPoiLon).toDouble();
+    const int lonM10 = qBound(firstLonM10, qFloor(lon * 10), lastLonM10);
+    // skip squares that were loaded before for this category
+    if (squaresToLoad.contains({categoryID, lonM10})) {
+      addPoi(query, categoryID, lonM10, minLatM10);
+    }
+  }
+}
+
+bool CPoiFilePOI::isLoaded(quint64 categoryID, int minLonM10, int minLatM10) const {
+  const auto category = loadedPoisByArea.constFind(categoryID);
+  if (category == loadedPoisByArea.constEnd()) {
+    return false;
+  }
+  const auto column = category->constFind(minLonM10);
+  return column != category->constEnd() && column->contains(minLatM10);
+}
+
+void CPoiFilePOI::addPoi(const QSqlQuery& query, quint64 categoryID, int minLonM10, int minLatM10) {
+  const quint64 key = query.value(CPoiFilePOIQuery::eSqlColumnPoiId).toULongLong();
+  const qreal lon = query.value(CPoiFilePOIQuery::eSqlColumnPoiLon).toDouble();
+  const qreal lat = query.value(CPoiFilePOIQuery::eSqlColumnPoiLat).toDouble();
+  const QStringList& data = query.value(CPoiFilePOIQuery::eSqlColumnPoiData).toString().split("\r");
+  QString garminIcon;
+  for (const QString& tag : data) {
+    if (tagMap.contains(tag)) {
+      garminIcon = tagMap[tag].getGarminSym();
+      break;
+    }
+  }
+  loadedPoisByArea[categoryID][minLonM10][minLatM10].append(key);
+  // TODO: this overwrites a POI if it already was loaded. The difference between those will be the category. Some
+  // better handling should be done
+  loadedPois[key] =
+      CPoiItemPOI(data, QPointF(lon * DEG_TO_RAD, lat * DEG_TO_RAD), key, categoryNames[categoryID], garminIcon);
 }
diff --git a/src/qmapshack/poi/CPoiFilePOI.h b/src/qmapshack/poi/CPoiFilePOI.h
index b75b18d8..fc596d07 100644
--- a/src/qmapshack/poi/CPoiFilePOI.h
+++ b/src/qmapshack/poi/CPoiFilePOI.h
@@ -24,10 +24,14 @@
 #include <QMutex>
 #include <QTimer>
 
+#include "poi/CPoiFilePOIQuery.h"
 #include "poi/CPoiIconCategory.h"
 #include "poi/CPoiItemPOI.h"
 #include "poi/IPoiFile.h"
 
+class QSqlDatabase;
+class QSqlQuery;
+
 class CPoiFilePOI : public IPoiFile {
   Q_DECLARE_TR_FUNCTIONS(CPoiFilePOI)
  public:
@@ -36,8 +40,8 @@ class CPoiFilePOI : public IPoiFile {
 
   void addTreeWidgetItems(QTreeWidget* widget) override;
   // category, minLon multiplied by 10, minLat multiplied by 10. POIs are loaded in squares of degrees (should be fine
-  // enough to not hang the system)
-  void loadPOIsFromFile(quint64 categoryID, int minLonM10, int minLatM10);
+  // enough to not hang the system). maxLonM10 is the last square of the row in view, used for files without R-tree.
+  void loadPOIsFromFile(quint64 categoryID, int minLonM10, int minLatM10, int maxLonM10);
 
   void draw(IDrawContext::buffer_t& buf) override;
 
@@ -63,23 +67,20 @@ class CPoiFilePOI : public IPoiFile {
     QSet<quint64> pois;
   };
 
-  enum SqlColumnPoi_e {
-    eSqlColumnPoiMaxLat,
-    eSqlColumnPoiMaxLon,
-    eSqlColumnPoiMinLat,
-    eSqlColumnPoiMinLon,
-    eSqlColumnPoiData,
-    eSqlColumnPoiId
-  };
   enum SqlColumnCategory_e { eSqlColumnCategoryId, eSqlColumnCategoryName, eSqlColumnCategoryParent };
 
   void getPoiIcon(QPixmap& icon, const poiGroup_t& poiGroup);
   void getPoiIcon(QPixmap& icon, const CPoiItemPOI& poi, const QString& definingTag = "");
   bool overlapsWithIcon(const QRectF& rect) const;
   bool getPoiGroupCloseBy(const QPoint& px, poiGroup_t& poiItem) const;
+  void loadPoisInSquare(const QSqlDatabase& db, quint64 categoryID, int minLonM10, int minLatM10);
+  void loadPoisInRow(const QSqlDatabase& db, int minLonM10, int maxLonM10, int minLatM10);
+  bool isLoaded(quint64 categoryID, int minLonM10, int minLatM10) const;
+  void addPoi(const QSqlQuery& query, quint64 categoryID, int minLonM10, int minLatM10);
 
   mutable QRecursiveMutex mutex;
   QString filename;
+  CPoiFilePOIQuery::Index index = CPoiFilePOIQuery::Index::eUnsupported;
   QTimer* loadTimer;
 
   QMap<quint64, Qt::CheckState> categoryActivated;
diff --git a/src/qmapshack/poi/CPoiFilePOIQuery.cpp b/src/qmapshack/poi/CPoiFilePOIQuery.cpp
new file mode 100644
index 00000000..abde6797
--- /dev/null
+++ b/src/qmapshack/poi/CPoiFilePOIQuery.cpp
@@ -0,0 +1,99 @@
+/**********************************************************************************************
+    Copyright (C) 2026 Gert Pellin <gert@pellin.be>
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+**********************************************************************************************/
+
+#include "poi/CPoiFilePOIQuery.h"
+
+#include <QSqlQuery>
+#include <QStringList>
+
+CPoiFilePOIQuery::Index CPoiFilePOIQuery::indexForVersion(const QString& version) {
+  if (version == "2" || version == "4") {
+    return Index::eRTree;
+  }
+  if (version == "3") {
+    return Index::eLatLon;
+  }
+  return Index::eUnsupported;
+}
+
+QString CPoiFilePOIQuery::selectPoisInSquare() {
+  return "SELECT (main.poi_index.minLat + main.poi_index.maxLat) / 2, "
+         "(main.poi_index.minLon + main.poi_index.maxLon) / 2, "
+         "main.poi_data.data, main.poi_data.id "
+         "FROM main.poi_data, main.poi_index "
+         "WHERE main.poi_data.id IN "
+         "("
+         "    SELECT main.poi_category_map.id "
+         "    FROM main.poi_category_map "
+         "    WHERE main.poi_category_map.id IN "
+         "    ( "
+         "        SELECT main.poi_index.id "
+         "        FROM main.poi_index "
+         "        WHERE main.poi_index.maxLat<:maxLat "
+         "        AND main.poi_index.minLat>=:minLat "
+         "        AND main.poi_index.maxLon<:maxLon "
+         "        AND main.poi_index.minLon>=:minLon "
+         "    ) "
+         "    AND main.poi_category_map.category=:categoryID "
+         ") "
+         "AND main.poi_data.id = main.poi_index.id";
+}
+
+void CPoiFilePOIQuery::bindSquare(QSqlQuery& query, quint64 categoryID, qint32 minLonM10, qint32 minLatM10) {
+  query.bindValue(":maxLat", (minLatM10 + 1) / 10.);
+  query.bindValue(":minLat", minLatM10 / 10.);
+  query.bindValue(":maxLon", (minLonM10 + 1) / 10.);
+  query.bindValue(":minLon", minLonM10 / 10.);
+  query.bindValue(":categoryID", categoryID);
+}
+
+QString CPoiFilePOIQuery::selectPoisInRow(qsizetype categoryCount) {
+  QStringList categories;
+  for (qsizetype i = 0; i < categoryCount; i++) {
+    categories << QString(":category%1").arg(i);
+  }
+
+  // CROSS JOIN fixes the order: walk the latitude index (it holds the id), look up the categories of each POI by
+  // its id, and read position and data of the matches only. The unary + keeps SQLite from using the longitude and
+  // category columns as index instead, which scans the whole file.
+  return QString(
+             "SELECT position.lat, position.lon, main.poi_data.data, main.poi_data.id, "
+             "main.poi_category_map.category "
+             "FROM main.poi_index AS band "
+             "CROSS JOIN main.poi_category_map ON main.poi_category_map.id = band.id "
+             "CROSS JOIN main.poi_index AS position ON position.id = band.id "
+             "CROSS JOIN main.poi_data ON main.poi_data.id = band.id "
+             "WHERE band.lat<:maxLat "
+             "AND band.lat>=:minLat "
+             "AND +main.poi_category_map.category IN (%1) "
+             "AND +position.lon<:maxLon "
+             "AND +position.lon>=:minLon")
+      .arg(categories.join(", "));
+}
+
+void CPoiFilePOIQuery::bindRow(QSqlQuery& query, const QList<quint64>& categoryIDs, qint32 minLonM10, qint32 maxLonM10,
+                               qint32 minLatM10) {
+  query.bindValue(":maxLat", (minLatM10 + 1) / 10.);
+  query.bindValue(":minLat", minLatM10 / 10.);
+  query.bindValue(":maxLon", (maxLonM10 + 1) / 10.);
+  query.bindValue(":minLon", minLonM10 / 10.);
+  for (qsizetype i = 0; i < categoryIDs.size(); i++) {
+    // a quint64 is bound as text, which the unary + leaves without the column's integer affinity
+    query.bindValue(QString(":category%1").arg(i), static_cast<qint64>(categoryIDs[i]));
+  }
+}
diff --git a/src/qmapshack/poi/CPoiFilePOIQuery.h b/src/qmapshack/poi/CPoiFilePOIQuery.h
new file mode 100644
index 00000000..7315fa5f
--- /dev/null
+++ b/src/qmapshack/poi/CPoiFilePOIQuery.h
@@ -0,0 +1,97 @@
+/**********************************************************************************************
+    Copyright (C) 2026 Gert Pellin <gert@pellin.be>
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+**********************************************************************************************/
+
+#ifndef CPOIFILEPOIQUERY_H
+#define CPOIFILEPOIQUERY_H
+
+#include <QList>
+#include <QString>
+
+class QSqlQuery;
+
+/**
+   @brief SQL for the POI position index of a mapsforge POI file
+
+   Version 2 and 4 keep the positions in an R-tree (minLat, maxLat, minLon, maxLon). Version 3 keeps them in a plain
+   table (lat, lon) with an index on each column. There a query has to visit every POI of a full latitude row,
+   whatever its longitude range, and look up its categories. So POIs are loaded by row segment and for several
+   categories at once. Everything else CPoiFilePOI reads is the same in all three versions.
+
+   Areas are given in squares of 0.1 degree, as multiples of 0.1 degree of their south west corner.
+ */
+class CPoiFilePOIQuery {
+ public:
+  enum class Index { eUnsupported, eRTree, eLatLon };
+
+  /// Result columns of selectPoisInSquare() and selectPoisInRow()
+  enum SqlColumnPoi_e {
+    eSqlColumnPoiLat,
+    eSqlColumnPoiLon,
+    eSqlColumnPoiData,
+    eSqlColumnPoiId,
+    eSqlColumnPoiCategory /**< selectPoisInRow() only */
+  };
+
+  /**
+     @brief Get the index layout for the value of the `version` metadata entry
+     @param version  the version as stored in the file
+     @return eUnsupported for any version that cannot be read
+   */
+  static Index indexForVersion(const QString& version);
+
+  /**
+     @brief Get the statement selecting all POIs of one category in one square, for Index::eRTree
+
+     Bind the values with bindSquare().
+
+     @return the SQL statement
+   */
+  static QString selectPoisInSquare();
+
+  /**
+     @brief Bind the values of the statement from selectPoisInSquare()
+     @param query       the prepared query
+     @param categoryID  the category of the POIs
+     @param minLonM10   longitude of the square
+     @param minLatM10   latitude of the square
+   */
+  static void bindSquare(QSqlQuery& query, quint64 categoryID, qint32 minLonM10, qint32 minLatM10);
+
+  /**
+     @brief Get the statement selecting all POIs of several categories in a row of squares, for Index::eLatLon
+
+     A POI in more than one of the categories is returned once per category. Bind the values with bindRow().
+
+     @param categoryCount  the number of categories, at least 1
+     @return the SQL statement
+   */
+  static QString selectPoisInRow(qsizetype categoryCount);
+
+  /**
+     @brief Bind the values of the statement from selectPoisInRow()
+     @param query        the prepared query
+     @param categoryIDs  the categories of the POIs, as many as passed to selectPoisInRow()
+     @param minLonM10    longitude of the first square of the row
+     @param maxLonM10    longitude of the last square of the row
+     @param minLatM10    latitude of the row
+   */
+  static void bindRow(QSqlQuery& query, const QList<quint64>& categoryIDs, qint32 minLonM10, qint32 maxLonM10,
+                      qint32 minLatM10);
+};
+
+#endif  // CPOIFILEPOIQUERY_H
diff --git a/src/qmapshack/poi/CPoiItemPOI.cpp b/src/qmapshack/poi/CPoiItemPOI.cpp
index d23efeb7..4f6c5292 100644
--- a/src/qmapshack/poi/CPoiItemPOI.cpp
+++ b/src/qmapshack/poi/CPoiItemPOI.cpp
@@ -42,8 +42,10 @@ CPoiItemPOI::CPoiItemPOI(const QStringList& data, const QPointF& coordinates, co
       if (keyValue[0].contains("wikidata", Qt::CaseInsensitive)) {
         wikidataRelatedKeys += keyValue[0];
       }
-      if (keyValue[0] == "name" || keyValue[0].contains("name:\\w\\w", Qt::CaseInsensitive) ||
-          keyValue[0].contains(re_match_name)) {
+      // normalized_name is the lowercase name without accents the mapsforge writer adds for searching
+      if ((keyValue[0] == "name" || keyValue[0].contains("name:\\w\\w", Qt::CaseInsensitive) ||
+           keyValue[0].contains(re_match_name)) &&
+          keyValue[0] != "normalized_name") {
         nameRelatedKeys += keyValue[0];
       }
       this->data[keyValue[0]] = keyValue[1];
@@ -183,7 +185,7 @@ QString CPoiItemPOI::getDesc() const {
   for (const QString& key : keys) {
     if (skipKeys.contains(key) || tempSkipKeys.contains(key) || wikidataRelatedKeys.contains(key) ||
         wikipediaRelatedKeys.contains(key) || nameRelatedKeys.contains(key) || key == "phone" ||
-        key == "contact:phone" || key == "contact:mobile" || key == "fax") {
+        key == "contact:phone" || key == "contact:mobile" || key == "fax" || key == "normalized_name") {
       continue;
     }
 
-- 
2.55.0

